cornell-dti/course-plan
# Entrance + Graduation Year Dropdown Revamp and Validation

### Summary 

This PR revamps the entrance year and graduation year dropdowns in Onboarding Basic Information. The dropdowns are based on the current year and the user's selections, ensuring that the user cannot select invalid years. This PR also validates the user's selection for seasons when the entrance and graduation year are the same.

- [x] populate entrance year dropdown based on current year
- [x] populate entrance year dropdown based on selected entrance year
- [x] suggest entrance year based on current year
- [x] suggest graduation year based on entrance year
- [x] clear graduation year selection if a new entrance year is after old grad year
- [x] when no seasons are selected, check that graduation year is at least entrance year
- [x] check that graduation semester is after entrance semester
- [x] introduce constants for dropdown size
- [x] use dynamic default grad year dropdown with no entrance year

### Test Plan 

- Open the preview URL with a new account, or delete your document for `user-onboarding-data` in Firestore.
- Navigate to "Profile".
- Check that the entrance year options are 6 +/- the current year (2022).
- Check that the graduation year options range from the current year (2022) to 12 years in the future (2032).
- Select an entrance year, and verify that the graduation year options now range from entrance year to entrance year + 8.
- Select a graduation year distinct from the entrance year and verify that your selection is considered valid.
- Select a graduation year = entrance year and verify that the seasons are compared correctly (graduation season ≥ entrance season).
- Select an entrance year that is after the current graduation year and ensure that the graduation year dropdown is cleared and that you are told to select a graduation year.

Also:
- Delete the value for `entranceSem` and `gradSem` for your document under `user-onboarding-data` in Firestore to simulate a user who created their account before seasons were added.
- Select entrance year = graduation year and verify that the only error message is "Please select an entrance semester and a graduation semester."

If desired, you can ensure that the entrance years are generated correctly by changing the return value for `getCurrentYear()` in [utilities.ts](src/utilities.ts) to return a different year (not 2022).

Note that there should never be an error message for the graduation semester being before the entrance semester when the seasons are blank because the graduation year should automatically be cleared when an entrance year that is after the graduation year is selected. The only time the user should receive and error for the graduation semester being before the entrance semester is when the years are the same and seasons are both not blank.

### Notes 

- I chose to allow a user to enter and graduate in the same semester (e.g. Fall 2021 and Fall 2021).
- The order of semesters in a year is: Winter, Spring, Summer, Fall.
- Depends on #647
